torch/csrc/DataLoader.cpp

```diff
@@ -0,0 +1,204 @@
+#include <sys/wait.h>
+#include <set>
+#include <atomic>
+#include <signal.h>
+#include "THP.h"
+
+// In cases like DataLoader, if a worker process die due to bus error/segfault
+// or just hang, the main process, if implemented with
+// multiprocessing.queue.SimpleQueue, will hang waiting for data. This is
+// difficult to avoid on PyTorch side as it can be caused by limited shm, or
+// other libraries users call in the workers. The following methods is an effort
+// to do our best provide some error message to users when such unfortunate
+// events happen.
+
+// TODO: The following don't work on Windows. Specifically, waitpid calls and
+// SIGCHLD handler.
+
+#ifndef _WIN32
+
+// Critical signal handlers should be registered on worker processes before
+// doing work.
+// Python handle is _set_worker_signal_handlers().
+#define SIGNAL_HANDLER(SIGNAL, HANDLER_NAME, ERROR_MSG)                       \
+static void HANDLER_NAME(int sig, siginfo_t *info, void *ctx)                 \
+{                                                                             \
+    write(STDERR_FILENO, ERROR_MSG, sizeof(ERROR_MSG) / sizeof(char));        \
+    _exit(EXIT_FAILURE);                                                      \
+}
+
+// signal(2) is really not portable. So use sigaction.
+// http://man7.org/linux/man-pages/man2/signal.2.html
+static int setSignalHandler(int signal, void(*handler)(int, siginfo_t *, void *), struct sigaction *old_sa_ptr)
+{
+  struct sigaction sa;
+  sa.sa_sigaction = handler;
+  sa.sa_flags = SA_RESTART|SA_SIGINFO|SA_NOCLDSTOP;
+  sigemptyset(&sa.sa_mask);
+  return sigaction(signal, &sa, old_sa_ptr);
+}
+
+SIGNAL_HANDLER(SIGBUS, handler_SIGBUS, "ERROR: Unexpected bus error encountered in worker. "
+  "This might be caused by insufficient shared memory (shm).\n");
+SIGNAL_HANDLER(SIGSEGV, handler_SIGSEGV, "ERROR: Unexpected segmentation fault encountered in worker.\n");
+
+PyObject *THPModule_setWorkerSignalHandlers(PyObject *module, PyObject *arg) {
+  HANDLE_TH_ERRORS
+  int error = 0;
+  error |= setSignalHandler(SIGBUS, &handler_SIGBUS, NULL) != 0;
+  error |= setSignalHandler(SIGSEGV, &handler_SIGSEGV, NULL) != 0;
+  return PyBool_FromLong(!error);
+  END_HANDLE_TH_ERRORS
+}
+
+static std::vector<pid_t> worker_pid_vec = {};
+// The following are needed since std::vector is not asynchronous safe.
+static std::atomic<pid_t *> worker_pids;
+static std::atomic<size_t> num_worker_pids(0);
+
+static void updatePIDsArray() {
+  size_t new_size =  worker_pid_vec.size();
+  auto new_ptr = (pid_t *)malloc(sizeof(pid_t) * new_size);
+  for (size_t idx = 0; idx < new_size; idx++) {
+    new_ptr[idx] = worker_pid_vec[idx];
+  }
+  pid_t *old_ptr = worker_pids;
+  if (new_size < num_worker_pids) {
+    num_worker_pids = new_size;
+    worker_pids = new_ptr;
+  } else {
+    worker_pids = new_ptr;
+    num_worker_pids = new_size;
+  }
+  free(old_ptr);
```

Comment:
I still don't see how this is supposed to be thread-safe or signal-safe. The signal handler could be called while this function is executing or it could be using the old_ptr which is freed here.

Reply:
If the handler is called before `free`, shouldn't it be using `worker_pids` instead? I am under the impression that signal handler can be called during this function but will suspend this function while executing. My thought is that this is thread-safe as it is guarded by GIL. Maybe I'm wrong..

Reply:
It may not be the same thread. I don't see any code that blocks the signal handler for other threads. Even if you did that, there's no guarantee that THPModule_setMainSignalHandlers is only called from a single thread.

Reply:
Well the signal can be handled in a thread that's different than the one running this function. Signal handlers can be parallel with parts of the program

Reply:
Thanks, I added mechanism to prevent update and handler to run in parallel. :)